---
title: "combined_data_prep"
output: html_document
date: "2023-09-20"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
setwd("~/project-gandalm/GandalLab")
output_dir <- ("~/project-gandalm/comb_data/processing")
plot_dir <- "/u/home/k/kafadare/project-gandalm/plots"
source("data_functions.R")
source("analysis_fcts.R")
# List of required packages
required_packages <- c("magrittr", "dplyr", "data.table", "DESeq2", "WGCNA", "sva", "ggplot2")
# Install or load missing packages
load_install_pkg(required_packages)
```


#Load data
```{r}
fetal_only_genExp <- read.table("/u/home/k/kafadare/project-gandalm/comb_data/new_data/fetal_only_genExp.tsv", header = T)
gtf_fetal <- read.table("/u/home/k/kafadare/project-gandalm/comb_data/new_data/gene_info_fetal.tsv", header = T)
#colnames(gtf_fetal)[2] <- "Chr"

adult_only_genExp <- read.table("/u/home/k/kafadare/project-gandalm/comb_data/new_data/adult_only_genExp.tsv", header = T)
gtf_adult <- read.table("/u/home/k/kafadare/project-gandalm/comb_data/new_data/gene_info_adult.tsv", header = T)
#colnames(gtf_adult)[2] <- "Chr"

comb_genExp.counts <- read.table("/u/home/k/kafadare/project-gandalm/comb_data/new_data/combo_genExp_counts.tsv", header = T)
comb_genExp.tpm <- read.table("/u/home/k/kafadare/project-gandalm/comb_data/new_data/combo_genExp_tpm.tsv", header = T)
gtf_comb <- read.table("/u/home/k/kafadare/project-gandalm/comb_data/new_data/gene_info_comb.tsv", header = T)
#colnames(gtf_comb)[2] <- "Chr"

combo_meta <- read.table("/u/home/k/kafadare/project-gandalm/comb_data/combo_meta.tsv", header = T)
##Change names in meta file to match the names in the genExp files (especially for the overlap samples)

```

``` {r}
fetal_counts_path <- "~/project-gandalm/fetal_rmlow_counts.tsv"
fetal_tpm_path <- "~/project-gandalm/fetal_rmlow_tpm.tsv"
#fetal_counts_path <- "/u/project/gandalm/cindywen/isoform_twas/salmon/expression.final/gene.noVersion.scaled.counts.tsv"
#fetal_tpm_path <- "/u/project/gandalm/cindywen/isoform_twas/salmon/expression.final/gene.noVersion.TPM.tsv"
fetal_meta_path <- "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/metadata_654.tsv"
#adult paths
adult_counts_path <- "~/project-gandalm/adult.counts.scaled.tsv"
adult_tpm_path <- "~/project-gandalm/adult.TPM.tsv"
adult_meta_path <- "/u/project/gandalm/kafadare/cov_hcp0_gene.txt"
#load data for fetal & adult
fetal <- load_data(names = c("counts", "tpm", "meta"), fetal_counts_path, fetal_tpm_path, fetal_meta_path)
adult <- load_data(names = c("counts", "tpm", "meta"), adult_counts_path, adult_tpm_path, adult_meta_path)
colnames(adult$counts) <- id_format_fix(colnames(adult$counts))
colnames(adult$tpm) <- id_format_fix(colnames(adult$tpm))

#get Chr number and start and end codon
gtf_fetal_all <- get_gene_info(fetal$counts$genid)
gtf_adult_all <- get_gene_info(adult$counts$genid)
```
Initial Histogram plots

```{r}
all_adult_plot <- prep_datExp(comb_genExp.counts,gtf_comb) %>% plot_gen_density(., title_str = "F/A Combined Counts")
all_fetal_plot <- prep_datExp(comb_genExp.tpm,gtf_comb) %>% plot_gen_density(., title_str = "F/A Combined Tpm")
all_adult_plot <- adult$counts %>% plot_gen_density(., title_str = "All Adult Counts")
all_fetal_plot <- fetal$counts %>% plot_gen_density(., title_str = "All Fetal Counts")
```
# 0. match with GTF file, remove chrX, chrY, chrM
# 1. round expression for DESeq2
# 2. lowly expressed removed already within each dataset
```{r}
##### 1
comb_counts_plot <- prep_datExp(comb_genExp.counts,gtf_comb) %>% plot_gen_density(., title_str = "F/A Combined Counts")
comb_tpm_plot <- prep_datExp(comb_genExp.tpm,gtf_comb) %>% plot_gen_density(., title_str = "F/A Combined Tpm")
fetal_counts_plot <- prep_datExp(fetal_only_genExp,gtf_fetal) %>% plot_gen_density(., title_str = "Fetal Only Counts", ylim = c(0, 1))
adult_counts_plot <- prep_datExp(adult_only_genExp,gtf_adult) %>% plot_gen_density(., title_str = "Adult Only Counts", ylim = c(0, 1))
```

# Visualization of Combat
```{r}
datExpr <- fread("/u/home/k/kafadare/project-gandalm/comb_data/new_data/combo_genExp_counts.tsv", header = T, data.table = F)
rownames(datExpr) <- datExpr$V1
datExpr <- datExpr[,-1]

datExpr.final <- fread("/u/home/k/kafadare/project-gandalm/comb_data/processed/counts.norm.noComBat.tsv", data.table = F)
rownames(datExpr.final) <- datExpr.final$V1
datExpr.final <- datExpr.final[,-1]

combat_expr <- fread("/u/home/k/kafadare/project-gandalm/comb_data/processed/counts.batch.processed.tsv", data.table = F)
rownames(combat_expr) <- combat_expr$V1
combat_expr <- combat_expr[,-1]

data.batch <- c()
  ##need to get data.batch
  for (i in 1:ncol(datExpr[,which(colnames(datExpr) %in% colnames(combat_expr))])) {
    sample <- colnames(datExpr[,which(colnames(datExpr) %in% colnames(combat_expr))])[i]
    #match the sample to id in meta file and find corresponding "study"
    data.batch[i] <- combo_meta[which(combo_meta$Subject %in% sample), "study"]
  }

s <- svd(datExpr[,which(colnames(datExpr) %in% colnames(combat_expr))] - rowMeans(datExpr[,which(colnames(datExpr) %in% colnames(combat_expr))]))
PC1 <- s$d[1]*s$v[,1]
PC2 <- s$d[2]*s$v[,2]
data <- data.frame(PC1, PC2, "Data" = factor(data.batch))
 
    
data.batch1 <- c()
  ##need to get data.batch
  for (i in 1:ncol(combat_expr)) {
    sample <- colnames(combat_expr)[i]
    #match the sample to id in meta file and find corresponding "study"
    data.batch1[i] <- combo_meta[which(combo_meta$Subject %in% sample), "study"]
  }

s_1 <- svd(combat_expr - rowMeans(combat_expr))
PC1_1 <- s_1$d[1]*s_1$v[,1]
PC2_1 <- s_1$d[2]*s_1$v[,2]
data1 <- data.frame(PC1_1, PC2_1, "Data" = factor(data.batch1))

#plot
options(repr.plot.width = 12, repr.plot.height = 6)

colnames(data1) <- c("PC1", "PC2", "Data")
data_combine <- rbind(data, data1)

data_combine$group <- c(rep("RawCounts", dim(data)[1]), 
                        rep("ComBat-seq", dim(data1)[1]))
data_combine$group <- factor(data_combine$group, levels = c("RawCounts", "ComBat-seq"))

ggplot(data_combine, aes(x = PC1, y = PC2, color = Data)) +
  geom_point(alpha = 0.5) +
  facet_wrap(~group, scales = "free") +
  labs(title = "Fetal Adult Combined Data, n = 1965", x = "", y = "") +
  theme_bw() +
  theme(axis.text = element_text(size = 16),
        plot.title = element_text(size = 18, face = "bold", hjust = 0.5),
        strip.text.x = element_text(size = 14),
        legend.title = element_text(size = 16, face = "bold"),
        legend.text = element_text(size = 14))


##the plot is not working - idk why >:(
# plot_norm <-  plot_gen_density(., title_str = "Norm & Batch Correct", ylim = c(0, 1))
# 
# 
# pdf(paste0(plot_dir,"ComBat_processed_shared.pdf"))
# plot_norm <-  plot_gen_density(combat_expr, title_str = "Norm & Batch Correct", ylim = c(0, 1))
# dev.off() 

```
