---
title: "combined_data_prep"
output: html_document
date: "2023-09-20"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
setwd("~/project-gandalm/GandalLab")
output_dir <- ("~/project-gandalm/comb_data/")
source("data_functions.R")
source("analysis_fcts.R")
# List of required packages
required_packages <- c("magrittr", "dplyr", "data.table", "DESeq2", "WGCNA", "sva", "ggplot2", "reshape2")
# Install or load missing packages
load_install_pkg(required_packages)
```

#Load data
```{r}
fetal_only_genExp <- read.table("/u/home/k/kafadare/project-gandalm/comb_data/fetal_only_genExp.tsv", header = T)
gtf_fetal <- read.table("/u/home/k/kafadare/project-gandalm/comb_data/gene_info_fetal.tsv", header = T)
colnames(gtf_fetal)[2] <- "Chr"
adult_only_genExp <- read.table("/u/home/k/kafadare/project-gandalm/comb_data/adult_only_genExp.tsv", header = T)
gtf_adult <- read.table("/u/home/k/kafadare/project-gandalm/comb_data/gene_info_adult.tsv", header = T)
colnames(gtf_adult)[2] <- "Chr"
comb_genExp.counts <- read.table("/u/home/k/kafadare/project-gandalm/comb_data/combo_genExp_counts.tsv", header = T)
comb_genExp.tpm <- read.table("/u/home/k/kafadare/project-gandalm/comb_data/combo_genExp_tpm.tsv", header = T)
gtf_comb <- read.table("/u/home/k/kafadare/project-gandalm/comb_data/gene_info_comb.tsv", header = T)
colnames(gtf_comb)[2] <- "Chr"
combo_meta <- read.table("/u/home/k/kafadare/project-gandalm/comb_data/combo_meta.tsv", header = T)
##Change names in meta file to match the names in the genExp files (especially for the overlap samples)
```

FUNCTIONS (will move to separate script later)

``` {r}
#prep data
prep_datExp <- function(.data, gtf){
gtf <- gtf %>% filter(!(Chr %in% c("chrX","chrY","chrM")))
.data <- .data %>% filter(genid %in% gtf$genid)
gtf <- gtf %>% filter(genid %in% .data$genid)
rownames(.data) <- .data$genid
.data <- .data[,!colnames(.data) %in% "genid"]
.data <- .data[,-1]
.data <- round(.data)
return(.data)
}

#function to plot gene density
plot_gen_density <- function(.data, offset = 0.1, title_str = 'Scaled Gene Counts ', xlim = c(-10,30), ylim = c(0,0.5), log = TRUE){
# View the distribution of expression for each sample.
# box plot, looking for big differences in read depth (raw counts), symmetry in distribution across samples
par(mfrow=c(1,2))
boxplot(.data, range = 0, main = paste(title_str), xlab = 'Samples', xaxt = "n")
boxplot(log2(offset+.data), range = 0, main = paste('log2(counts+',offset,')'), xlab = 'Samples', xaxt = "n")
# Histogram/density plot
# Look for: how well do the distributions line up, outlier samples, zero counts
par(mfrow=c(1,1))
i <- 1
  if(log == TRUE){
   plot <- plot(density(log2(offset+.data[,i])), main = paste('Scaled Gene    Counts ',title_str), xlab = paste('log2(counts+',offset,')'), 
        xlim = xlim, ylim = ylim)
     for(i in 1:ncol(.data)){
       lines(density(log2(.1+.data[,i])), col = i)
     }
  }else{
  plot <- plot(density(.data[,i]), main = paste('Scaled Gene Counts ',title_str),   xlab = 'Counts', 
       xlim = xlim, ylim = ylim)
    for(i in 1:ncol(.data)){
      lines(density(.data[,i]), col = i)
    }
  }
plot
return(plot)
}

#function to remove lowly expressed genes
rm_low <- function(datExpr,datExpr.tpm, gtf, cutoff = 0.1, percent = 0.25){
#remove unwanted Chr
#remove lowly expressed
# cutoff default 0.1, default 25% subjects
datExpr.tpm <- datExpr.tpm %>% filter(genid %in% gtf$genid)
rownames(datExpr.tpm) <- datExpr.tpm$genid
datExpr.tpm <- datExpr.tpm[,-1]
# cutoff=0.1, 25% subjects
keep <- (rowSums(datExpr.tpm > cutoff)) > percent*ncol(datExpr.tpm)
# # Count filter, use cpm, not raw or scaled from TPM counts!
# keep <- filterByExpr(datExpr, min.count = (cutoff*100), min.prop=percent)
datExpr <- datExpr[keep,]
return(datExpr)
}

# melted_data <- melt(comb_genExp.counts[,2:ncol(comb_genExp.counts)])
# ggplot(melted_data, aes(x = value)) +
#   geom_histogram(binwidth = 1, fill = "lightblue", color = "black") +
#   labs(title = "Gene Expression Density", x = "Expression Value", y = "Density") + theme_minimal()
```
Initial Histogram plots

```{r}

comb_counts_plot <- prep_datExp(comb_genExp.counts,gtf_comb) %>% plot_gen_density(., title_str = "F/A Combined Counts")
comb_tpm_plot <- prep_datExp(comb_genExp.tpm,gtf_comb) %>% plot_gen_density(., title_str = "F/A Combined Tpm")
fetal_counts_plot <- prep_datExp(fetal_only_genExp,gtf_fetal) %>% plot_gen_density(., title_str = "Fetal Only Counts", ylim = c(0, 1))
adult_counts_plot <- prep_datExp(adult_only_genExp,gtf_adult) %>% plot_gen_density(., title_str = "Adult Only Counts", ylim = c(0, 0.3))

```
# 0. match with GTF file, remove chrX, chrY, chrM
# 1. round expression for DESeq2
# 2. remove lowly expressed based on TPM
```{r}
##### 1
#datExpr <- fread("/u/project/gandalm/cindywen/isoform_twas/salmon/expression.final/gene.noVersion.scaled.counts.tsv", data.table = F)
#gtf_fetal <- fread("/u/project/gandalm/cindywen/isoform_twas/salmon/gencode.gene.noVersion.tsv", data.table = F)
#fetal_tpm <- fread("/u/project/gandalm/cindywen/isoform_twas/salmon/expression.final/gene.noVersion.TPM.tsv", data.table = F)
#fetal_tpm <- fetal[[1]]
#fetal_tpm$genid <- rownames(fetal_tpm)
#colnames(fetal_gtf) <- c("Chr", "V2", "V3", "V4", "genid")
#colnames(fetal_tpm)[1] <- "genid"
#walker and hdbr duplicate 1707
#make hdbr 1707.1
#names(datExpr)[647] <- "1707.1"

keep_fetal <- rm_low(fetal_tpm, gtf_fetal, 0.1, 0.25)

keep_shared <- rm_low(comb_genExp.counts, comb_genExp.tpm, gtf_comb, 0.1, 0.25)
plots <- list()
plots[1] <- prep_datExp(keep_shared, gtf_comb) %>% plot_gen_density(., title_str = "Low RM 0.1/0.25", ylim = c(0, 0.5))
plots[2] <- prep_datExp(keep_shared, gtf_comb) %>% plot_gen_density(., title_str = "Low RM 0.5/0.25", ylim = c(0, 0.5))
plots[3] <- prep_datExp(keep_shared, gtf_comb) %>% plot_gen_density(., title_str = "Low RM 0.5/0.5", ylim = c(0, 0.5))

```

# 3. VST normalization and transformation
# 4. remove outlier subjects by connectivity
# 5. ComBat
```{r}

norm_batch <- function(datExpr, meta){
datExpr.vst <- varianceStabilizingTransformation(as.matrix(datExpr), blind = TRUE)
datExpr.vst <- as.data.frame(datExpr.vst)
##### 3
# Normalize
datExpr.vst <- varianceStabilizingTransformation(as.matrix(datExpr), blind = TRUE)
datExpr.vst <- as.data.frame(datExpr.vst)
#### 4
# Remove outliers
normadj <- adjacency(datExpr.vst,type = 'signed',corFnc = 'bicor')   #Calculate network adjacency
netsummary <- fundamentalNetworkConcepts(normadj)
C <- netsummary$Connectivity   #Extract connectivity of each sample
Z.C <- (C-mean(C))/sqrt(var(C))   #Covert to Z-score
outliers <- (Z.C < -3)
datExpr.final <- datExpr.vst
datExpr.final <- datExpr.final[,!outliers]
write.table(datExpr.final, paste0(args$outdir,args$level,".counts.processed.noComBat.tsv"), col.names = T, row.names = T, quote = F, sep = "\t")
exprMat <- as.matrix(datExpr.final)
# ComBat
##need to get data.batch
# for (i in 1:ncol(datExpr.final)) {
#   sample <- colnames(datExpr.final)[i]
#   #match the sample to id in meta file and find corresponding "study"
#     data.batch[i] <- meta[which(meta$Subject %in% sample), "study"]
# }
combat_expr <- ComBat(dat = exprMat, batch = data.batch, mod = NULL, par.prior = TRUE, prior.plots = FALSE)
combat_expr <- as.data.frame(combat_expr)

write.table(combat_expr, paste0(args$outdir, args$level, ".counts.processed.tsv"), col.names = T, row.names = T, quote = F, sep = "\t")

outlier.df <- data.frame(c(which(outliers)))
outlier.df$c.which.outliers.. <- rownames(outlier.df)
write.table(outlier.df,paste0(args$outdir, args$level, ".outlier.txt"), sep="\t", quote=F, col.names = F, row.names = F)
return(datExpr)
}

norm_shared <- norm_batch(keep_shared, combo_meta)
plots[1] <- prep_datExp(keep_shared, gtf_comb) %>% plot_gen_density(., title_str = "Low RM 0.1/0.25", ylim = c(0, 0.5))
plots[2] <- prep_datExp(keep_shared, gtf_comb) %>% plot_gen_density(., title_str = "Low RM 0.5/0.25", ylim = c(0, 0.5))
plots[3] <- prep_datExp(keep_shared, gtf_comb) %>% plot_gen_density(., title_str = "Low RM 0.5/0.5", ylim = c(0, 0.5))

```

# 6. pheno.bed for FastQTL; note strand; note 0/1-based

```{r}
##### 6
# phenotype.bed
gtf <- gtf %>% filter(V5 %in% rownames(datExpr.final))

# prepare BED file
setDT(combat_expr, keep.rownames = TRUE)
bed <- merge(combat_expr, gtf, by.x = "rn", by.y = "V5", all = TRUE)
bed <- as.data.frame(bed)

bed <- bed[, c(ncol(bed)-3, ncol(bed)-2, ncol(bed)-1, ncol(bed), 1:(ncol(bed)-4))]

# special note! strand; 0/1-based conversion
bed$V1 <- gsub("chr","",bed$V1)

for (i in 1:dim(bed)[1]){
	if (bed[i,4] == "+") {
		gtf_start=bed[i,2]
		bed[i,2]=gtf_start-1
		bed[i,3]=gtf_start
	}
	else if (bed[i,4] == "-") {
		gtf_end=bed[i,3]
		bed[i,2]=gtf_end-1
		bed[i,3]=gtf_end
	}
}

bed <- bed[order(bed$V1, bed$V2),]

bed <- bed %>% select(-V4)
colnames(bed)[1:4] <- c("#Chr", "start", "end", "ID")

write.table(bed, paste0(args$outdir, args$level, ".counts.scaled.normalized.bed"), quote=F, sep="\t", row.names=F, col.names=T)
```